apiserver/facades/client/spaces/spaces.go

```diff
@@ -369,22 +376,16 @@ func (api *API) ensureSpacesAreMutable(ctx stdcontext.Context) error {
 // An error is returned if it is the provider and not the Juju operator
 // that determines the space topology.
 func (api *API) ensureSpacesNotProviderSourced(ctx stdcontext.Context) error {
```

Comment:
I think we have a bug here that we are going to keep perpetuating with this change.

This func says:
```
// ensureSpacesNotProviderSourced checks if the environment implements
// NetworkingEnviron and also if it supports provider spaces.
// An error is returned if it is the provider and not the Juju operator
// that determines the space topology.
```

We know that if `providerGetter` returns a `NotSupported` error that provider spaces are not supported and so so the func is satisfied true but we return an error. Seems like our logic is broken from previous implementation.

Can we do some research on if this is correct or not, check what tests we have around this and let's look to fix this properly. May also be worth getting Joe's two cents but at the moment this looks to be a logic bug we are just moving.

Reply:
Re: testing, this function is tested indirectly in the `TestRenameSpace*` and `TestRemoveSpace*` tests.

Reply:
I'm struggling to wrap my head around what this function should be doing, so it's hard for me to tell if it's correct or not.

Basically there are 4-5 different cases we have to handle, so we need to work out what we should be doing here in each of these cases:
- provider doesn't support networking
- provider.SupportsSpaceDiscovery returns an error
- provider.SupportsSpaceDiscovery returns false
- provider.SupportsSpaceDiscovery returns true

We also have to worry about what we should be doing in the case where the `credentialInvalidatorGetter` function returns an error.

Reply:
Have had a look at this and it is fine to leave at the moment.

Reply:
Looking again at this function, basically it returns nil if and only if the provider has a `SupportsSpaceDiscovery` method and that method returns false.

If the providerGetter returns a NotSupported error, then the environment doesn't support networking and we error out. I think that is fine as we should not be supporting any operations on spaces if the env doesn't support networking.